airbnb_project/listings/harvester_app/harvester/spiders/listings_spider.py

```diff
@@ -0,0 +1,312 @@
+ZOOM_LEVEL = 15.4
+
+
+
+def n(s):
+    """
+    Encode the input string `s` into bytes, then encode it in base64, and decode it back to a UTF-8 string.
+
+    :param s: The input string to be encoded.
+    :return: The base64 encoded string.
+    """
+    return base64.b64encode(s.encode('utf-8')).decode('utf-8')
+
+
+from urllib.parse import quote
+
+
+def t(n, t):
+    """
+    Encode the second string `t` using URL encoding, replace spaces with plus signs, and return the combined format.
+
+    :param n: The first part of the string to be combined.
+    :param t: The string to be URL encoded and combined.
+    :return: A formatted string combining `n` and the encoded version of `t`.
+    """
+    encoded_t = quote(t, safe='').replace('%20', '+').replace('(', '%28').replace(')', '%29')
+    return f"{n}:{encoded_t}"
+
+
+class ListingsSpider(scrapy.Spider):
+    """
+    Spider for scraping Airbnb listings.
+    """
+    # Name of the spider, used by Scrapy to identify this spider
+    name = "listings_spider"
+
+    # List of allowed domains for the spider to crawl
+    allowed_domains = ['airbnb.ca']
+
+    # List to store pagination cursors for navigating through result pages
+    next_page_cursors: [str] = None
+
+    # Counter to keep track of the total number of listings scraped
+    total_listings = 0
+
+    # Set to store IDs of listings that have already been scraped to avoid duplicates. This is a
+    # temporary solution until a database is set up, which will then take care of duplicates etc
+    ALREADY_SCRAPED_LISTINGS = set()
+
+    # Instance of AirBnbURLBuilder to generate URLs for different search parameters
+    URL_BUILDER = AirBnbURLBuilder()
+
+    def start_requests(self):
+        """
+       Generates initial requests for scraping Airbnb listings using coordinates from a JSON file.
+
+       This function performs the following steps:
+       1. Reads already scraped Airbnb listing IDs from a CSV file and stores them in a set.
+       2. Reads coordinates from a JSON file and shuffles them.
+       3. Constructs and returns a list of initial Scrapy requests using template URLs and the read coordinates.
+
+       :return: List of Scrapy FormRequest objects initialized with map coordinates for Vancouver.
+        """
+        file_name = self.settings.get("CSV_STORE_FILE_NAME")
+        try:
+            with open(
+                    f"{file_name}",
+                    "r",
+                    encoding="utf8") as file:
+                read_csv_file = csv.DictReader(file)
+                for row in read_csv_file:
+                    row: dict
+                    ListingsSpider.ALREADY_SCRAPED_LISTINGS.add(row['airbnb_listing_id'])
+        except FileNotFoundError as e:
+            print(e)
+        requests = []
+        try:
+            with open('cordinates.json', 'r') as file:
+                coordinates = json.load(file)
+                urls = ListingsSpider.URL_BUILDER.get_urls("july", "june", "august")
```

Comment:
does this mean only results from June to August? should there be a function to customize it instead?